pkg/controllers/vas/subclusterresize_reconcile.go

```diff
@@ -0,0 +1,112 @@
+/*
+ (c) Copyright [2021-2022] Micro Focus or one of its affiliates.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ You may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package vascontroller
+
+import (
+	"context"
+
+	vapi "github.com/vertica/vertica-kubernetes/api/v1beta1"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	verrors "github.com/vertica/vertica-kubernetes/pkg/errors"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	"github.com/vertica/vertica-kubernetes/pkg/vasstatus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// SubclusterResizeReconciler will grow/shrink existing subclusters based on the
+// target pod count in the CR.
+type SubclusterResizeReconciler struct {
+	VRec *VerticaAutoscalerReconciler
+	Vas  *vapi.VerticaAutoscaler
+	Vdb  *vapi.VerticaDB
+}
+
+func MakeSubclusterResizeReconciler(r *VerticaAutoscalerReconciler, vas *vapi.VerticaAutoscaler) controllers.ReconcileActor {
+	return &SubclusterResizeReconciler{VRec: r, Vas: vas, Vdb: &vapi.VerticaDB{}}
+}
+
+// Reconcile will grow/shrink an existing subcluste based on the target pod count
+func (s *SubclusterResizeReconciler) Reconcile(ctx context.Context, req *ctrl.Request) (ctrl.Result, error) {
+	if s.Vas.Spec.ScalingGranularity != vapi.PodScalingGranularity {
+		return ctrl.Result{}, nil
+	}
+
+	return s.resizeSubcluster(ctx, req)
+}
+
+// resizeSubcluster will change the size of a subcluster given the target pod count
+func (s *SubclusterResizeReconciler) resizeSubcluster(ctx context.Context, req *ctrl.Request) (ctrl.Result, error) {
+	var res ctrl.Result
+	scalingDone := false
+	// Update the VerticaDB with a retry mechanism for any conflict updates
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```

Comment:
It handles the case if someone else had update the object since you last fetched it.  This will attempt to retry the update again by fetching the more recent copy of the object.